sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/cardinality/ApproximateDistinct.java

```diff
@@ -0,0 +1,352 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@code PTransform}s for computing the number of distinct elements in a {@code PCollection}, or
+ * the number of distinct values associated with each key in a {@code PCollection} of {@code KV}s.
+ *
+ * <p>This class uses the HyperLogLog algorithm, and more precisely
+ * the improved version of google (HyperLogLog+).
+ * <br>The implementation comes from Addthis' library Stream-lib : https://github.com/addthis/stream-lib
+ * <br>The original paper of the HyperLogLog is available here : http://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf
+ * <br>A paper from the same authors to have a clearer view of the algorithm is available here :
+ * http://cscubs.cs.uni-bonn.de/2016/proceedings/paper-03.pdf
+ * <br>Google's HyperLogLog+ version is detailed in this paper : https://research.google.com/pubs/pub40671.html
+ */
+public class ApproximateDistinct {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ApproximateDistinct.class);
+
+    // do not instantiate
+  private ApproximateDistinct() {
+  }
+
+  /**
+   * A {@code PTransform} that takes an inputT {@code PCollection} of objects and returns a
+   * {@code PCollection<HyperLogLogPlus>} whose contents is a sketch which approximates
+   * the number of distinct element in the input {@code PCollection}.
+   *
+   * <p>The {@code p} parameter controls the accuracy of the estimation. It represents
+   * the number of bits that will be used to index the elements,
+   * thus the number of different "buckets" in the HyperLogLog+ sketch.
+   * <br>In general, you can expect a relative error of about :
+   * <pre>{@code 1.1 / sqrt(2^p)}</pre>
+   * For instance, the estimation {@code ApproximateDistinct.globally(12)}
+   * will have a relative error of about 2%.
+   * <br> Also keep in mind that {@code p} cannot be lower than 4,
+   * because the estimation would be too inaccurate.
+   * <br>See {@link ApproximateDistinctFn} for more details about the algorithm's principle.
+   *
+   * <p>HyperLogLog+ version of Google uses a sparse representation in order to
+   * optimize memory and improve accuracy for small cardinalities.
+   * By calling this builder, you will not use the sparse representation.
+   * If you want to, see {@link ApproximateDistinctFn#withSparseRepresentation(int)}
+   *
+   * <p>Example of use
+   * <pre> {@code PCollection<String> input = ...;
+   * PCollection<HyperLogLogPlus> hllSketch = input
+   *        .apply(ApproximateDistinct.<String>globally(15));
+   * } </pre>
+   *
+   * @param <InputT>    type of elements being combined
+   * @param p           number of bits for indexes in the HyperLogLogPlus
+   *
+   *
+   */
+  public static <InputT> Combine.Globally<InputT, HyperLogLogPlus> globally(int p) {
+    return Combine.<InputT, HyperLogLogPlus>globally(ApproximateDistinctFn.<InputT>create(p));
+  }
+
+    /**
+     * Do the same as {@link ApproximateDistinct#globally(int)}, but with a default value for p.
```

Comment:
<!--new_thread; commit:6bf51fdf6afe6f389415792f335dd08adf9fa85a; resolved:0-->
`with a default value for p -> with a default value of 18 for p`